Classes/Controller/KrakenController.php

```diff
@@ -56,4 +72,46 @@ public function replaceLocalFileAction()
 
         $this->resourceService->replaceLocalFile($krakenIoResult);
     }
+
+    /**
+     * Replaces the local file within the file system with the optimized image delivered by Kraken.
+     *
+     * @throws \Neos\Flow\Exception
+     * @see https://kraken.io/docs/wait-callback
+     */
+    public function replaceThumbnailResourceAction()
+    {
+        $krakenIoResult = $this->request->getArguments();
+
+        if (!isset($krakenIoResult['success']) || $krakenIoResult['success'] !== 'true') {
+            throw new \Neos\Flow\Exception('Kraken was unable to optimize resource', 1524665608);
+        }
+
+        if (!isset($krakenIoResult['file_name'])) {
+            throw new \Neos\Flow\Exception('Filename missing in Kraken callback payload', 1524665605);
+        }
+
+        if (!isset($krakenIoResult['verificationToken']) ||
+            $this->krakenService->verifyToken($krakenIoResult['verificationToken'], $krakenIoResult['originalFilename']) === false) {
+            throw new \Neos\Flow\Exception('Invalid verification token supplied', 1524665601);
+        }
+
+        // Get thumbnail identifier
+        $resourceIdentifier = $krakenIoResult['resourceIdentifier'];
+        $query = $this->entityManager->createQuery('SELECT t FROM Neos\Media\Domain\Model\Thumbnail t WHERE t.resource = :resource');
+        $query->setParameter('resource', $resourceIdentifier);
+        $query->setMaxResults(1);
```

Comment:
We should take all thumbnails into account that use the resource

Reply:
Have it on my screen already. Thanks for the remind!